server/src/main/java/org/opensearch/indices/replication/OngoingSegmentReplications.java

```diff
@@ -0,0 +1,229 @@
+     * @param listener {@link ActionListener} that resolves when sending files is complete.
+     */
+    void startSegmentCopy(GetSegmentFilesRequest request, ActionListener<GetSegmentFilesResponse> listener) {
+        final DiscoveryNode node = request.getTargetNode();
+        final SegmentReplicationSourceHandler handler = nodesToHandlers.get(node);
+        if (handler != null) {
+            if (handler.isActive()) {
+                throw new OpenSearchException(
+                    "Replication to shard {}, on node {} has already started",
+                    request.getCheckpoint().getShardId(),
+                    request.getTargetNode()
+                );
+            }
+            // update the given listener to release the CopyState before it resolves.
+            final ActionListener<GetSegmentFilesResponse> wrappedListener = ActionListener.runBefore(listener, () -> {
+                final SegmentReplicationSourceHandler sourceHandler = nodesToHandlers.remove(node);
+                if (sourceHandler != null) {
+                    removeCopyState(sourceHandler.getCopyState());
+                }
+            });
+            handler.sendFiles(request, wrappedListener);
+        } else {
+            listener.onResponse(new GetSegmentFilesResponse(Collections.emptyList()));
+        }
+    }
+
+    /**
+     * Cancel any ongoing replications for a given {@link DiscoveryNode}
+     *
+     * @param node {@link DiscoveryNode} node for which to cancel replication events.
+     */
+    void cancelReplication(DiscoveryNode node) {
+        final SegmentReplicationSourceHandler handler = nodesToHandlers.remove(node);
+        if (handler != null) {
+            handler.cancel("Cancel on node left");
+            removeCopyState(handler.getCopyState());
+        }
+    }
+
+    /**
+     * Prepare for a Replication event. This method constructs a {@link CopyState} holding files to be sent off of the current
+     * nodes's store.  This state is intended to be sent back to Replicas before copy is initiated so the replica can perform a diff against its
+     * local store.  It will then build a handler to orchestrate the segment copy that will be stored locally and started on a subsequent request from replicas
+     * with the list of required files.
+     *
+     * @param request         {@link CheckpointInfoRequest}
+     * @param fileChunkWriter {@link FileChunkWriter} writer to handle sending files over the transport layer.
+     * @return {@link CopyState} the built CopyState for this replication event.
+     * @throws IOException - When there is an IO error building CopyState.
+     */
+    CopyState prepareForReplication(CheckpointInfoRequest request, FileChunkWriter fileChunkWriter) throws IOException {
+        final CopyState copyState = getCachedCopyState(request.getCheckpoint());
+        if (nodesToHandlers.containsKey(request.getTargetNode())) {
+            throw new OpenSearchException(
```

Comment:
:+1:

Reply:
computeIfAbsent actually returns the computed value, not the key.  For the behavior we want here I can use putIfAbsent and precompute the object.  creating the handlers is not expensive so I think this is ok.

```java
        if (nodesToHandlers.putIfAbsent(request.getTargetNode(), createTargetHandler(request.getTargetNode(), copyState, fileChunkWriter)) != null) {
            throw new OpenSearchException(
                "Shard copy {} on node {} already replicating",
                request.getCheckpoint().getShardId(),
                request.getTargetNode()
            );
        
```

Reply:
you are very right, thanks @mch2